klippy/extras/mixing_extruder.py

```diff
@@ -0,0 +1,335 @@
+    idx = idx - 1
+    if idx < steppers_count:
+        return [1. if p == idx else 0. for p in range(steppers_count)]
+    idx = idx - steppers_count
+    if idx < steppers_count:
+        return [0. if p == idx else 1. / (steppers_count - 1)
+                for p in range(steppers_count)]
+    idx = idx - steppers_count
+    if steppers_count == 3:
+        if idx < 2 * steppers_count:
+            return [[0. if p == x else (1 + (((p - x) % 3 + y) % 2)) / 3.
+                     for p in range(steppers_count)]
+                    for x in range(steppers_count) for y in (0, 1)][idx]
+        idx = idx - 2 * steppers_count
+    elif steppers_count > 3:
+        if idx < (steppers_count * (steppers_count - 1) / 2):
+            return [[0. if p == x or p == y else 1. / (steppers_count - 2)
+                     for p in range(steppers_count)]
+                    for x in range(steppers_count)
+                    for y in range(x + 1, steppers_count)][idx]
+    return [1. / steppers_count for p in range(steppers_count)]
+
+
+class MixingExtruder(object):
+    def __init__(self, config, parent=None):
+        self.printer = config.get_printer()
+        self._activated = False
+        self.name = "mixingextruder"
+        self.stepper_names = tuple(e.strip()
+                               for e in config.get('steppers', '').split(",")
+                               if len(e))
+        if not len(self.stepper_names):
+            raise self.printer.config_error(
+                "No steppers configured for mixing")
+        self._extruder_steppers = []
+        self._rotation_distances = ()
+        self._trapqs = ()
+        self.extruder_name = config.get('extruder_name',
+                                        self.stepper_names[0])
+        self._mixing = self._active_mixing = \
+            mixing_preset(0, len(self.stepper_names))
+        self._enabled = False
+
+        self.printer.register_event_handler("klippy:connect",
+                                            self._handle_connect)
+        # Register commands
+        gcode = self.printer.lookup_object('gcode')
+        gcode.register_command("SET_MIXING_EXTRUDER",
+                               self.cmd_SET_MIXING_EXTRUDER,
+                               desc=self.cmd_MIXING_STATUS_help)
+        gcode.register_command("MIXING_STATUS",
+                               self.cmd_MIXING_STATUS,
+                               desc=self.cmd_MIXING_STATUS_help)
+    def _handle_connect(self):
+        if self._activated:
+            return
+        self._activated = True
+        try:
+            for name in self.stepper_names:
+                try:
+                    stepper = (self.printer
+                               .lookup_object(name).extruder_stepper)
+                except:
+                    stepper = (self.printer
+                               .lookup_object("extruder_stepper " + name)
+                               .extruder_stepper)
```

Comment:
This doesn't look a good idea to me.  If the user wants to use an "extruder_stepper mystepper" then they should specify that explicitly in the config file.  I don't think the code should try to infer what stepper is being requested.

Reply:
I have adjusted it to use if/else instead of try since the "extruder" is actually fixed in name